File: android/app/src/main/java/com/noop/ui/BodyClockDialCard.kt
package com.noop.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.noop.R
import com.noop.analytics.CircadianEngine
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin

// MARK: - 24 h body-clock dial (#1680)
//
// Two concentric arcs on one ring: the outer is where the body clock wanted the night, the inner is where
// it happened. Overlap is the whole message — the card exists so "was last night's timing right" is
// answerable at a glance, which the text-only BodyClockCard on the Health screen cannot do.
//
// VOCABULARY. The caption measures sleepWindowOffsetHours — the distance between the two ARCS DRAWN — and
// NOT offsetVsScheduleMinutes, which compares the clock to the wearer's habitual schedule and is what the
// Health card already reports. The two disagree exactly when someone keeps a consistent schedule that does
// not suit their clock, and that is the case this dial exists to show, so captioning it with the other
// number would contradict the picture.
//
// Nothing here computes a metric: the window, the offset and the chronotype all come from CircadianEngine,
// byte-identical with the Swift twin. Only the drawing is per-platform (visual parity, not pixel parity).

/**
 * The dial card. [actualBedHour] / [actualWakeHour] are the night's own clock hours (0..<24, fractional),
 * taken from the scored session by the caller. Twin of Apple `BodyClockDialCard`.
 */
@Composable
fun BodyClockDialCard(
    estimate: CircadianEngine.PhaseEstimate,
    actualBedHour: Double,
    actualWakeHour: Double,
) {
    val hue = Palette.restColor
    // The night's length, taken the long way round the clock when it crosses midnight.
    val durationHours = ((actualWakeHour - actualBedHour) % 24.0).let { if (it <= 0.0) it + 24.0 else it }
    // null exactly when the night's length is non-positive or a full day — the same input that makes
    // sweepHours wrap to 24 h and draw the actual arc as a complete ring. A full circle with no ideal arc
    // beside it would state something false about the night, so the card stands down. Mirrors Swift.
    val ideal = CircadianEngine.idealSleepWindow(estimate.tempMinHour, durationHours) ?: return
    val offsetHours = CircadianEngine.sleepWindowOffsetHours(estimate.tempMinHour, actualWakeHour)
    val alignment = alignmentText(offsetHours)
    val dialLabel = uiString(R.string.l10n_body_clock_dial_card_body_clock_dial_03cdbc31)

    NoopCard(tint = hue) {
        Column(verticalArrangement = Arrangement.spacedBy(Metrics.gap)) {
            Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.Top) {
                Column(modifier = Modifier.weight(1f)) {
                    Overline(uiString(R.string.l10n_body_clock_dial_card_body_clock_b0b9b988))
                    Text(
                        uiString(R.string.l10n_body_clock_dial_card_last_night_against_your_clock_9183a37c),
                        style = NoopType.footnote,
                        color = Palette.textTertiary,
                    )
                }
            }

            Canvas(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(170.dp)
                    .semantics { contentDescription = dialLabel },
            ) {
                val side = min(size.width, size.height)
                val centre = Offset(size.width / 2f, size.height / 2f)
                val outer = side / 2f - 10.dp.toPx()
                val inner = outer - 16.dp.toPx()

                // The bare ring, so an empty dial still reads as a clock rather than a broken chart.
                drawCircle(color = Palette.hairline, radius = outer, center = centre,
                    style = Stroke(width = 1.dp.toPx()))

                // Six-hourly ticks: enough to orient the eye, few enough not to compete with the arcs.
                var tick = 0.0
                while (tick < 24.0) {
                    val a = Math.toRadians(hourAngleDegrees(tick))
                    drawLine(
                        color = Palette.textTertiary.copy(alpha = 0.5f),
                        start = Offset(centre.x + (cos(a) * (outer - 4.dp.toPx())).toFloat(),
                            centre.y + (sin(a) * (outer - 4.dp.toPx())).toFloat()),
                        end = Offset(centre.x + (cos(a) * outer).toFloat(),
                            centre.y + (sin(a) * outer).toFloat()),
                        strokeWidth = 1.dp.toPx(),
                    )
                    tick += 6.0
                }

                fun arc(radius: Float, from: Double, to: Double, colour: androidx.compose.ui.graphics.Color, width: Float) {
                    drawArc(
                        color = colour,
                        startAngle = hourAngleDegrees(from).toFloat(),
                        sweepAngle = (sweepHours(from, to) / 24.0 * 360.0).toFloat(),
                        useCenter = false,
                        topLeft = Offset(centre.x - radius, centre.y - radius),
                        size = Size(radius * 2, radius * 2),
                        style = Stroke(width = width, cap = StrokeCap.Round),
                    )
                }

                if (ideal != null) {
                    arc(outer, ideal.bedHour, ideal.wakeHour, hue.copy(alpha = 0.35f), 8.dp.toPx())
                }
                arc(inner, actualBedHour, actualWakeHour, hue, 8.dp.toPx())
            }

            Text(alignment, style = NoopType.title2, color = Palette.textPrimary)

            CircadianEngine.chronotype(estimate)?.let { c ->
                Text(chronotypeText(c), style = NoopType.footnote, color = Palette.textTertiary)
            }
        }
    }
}

/** A unix second as a fractional LOCAL clock hour — the dial's only input beyond the phase estimate. */
internal fun localClockHour(ts: Long): Double {
    val c = java.util.Calendar.getInstance()
    c.timeInMillis = ts * 1000L
    return c.get(java.util.Calendar.HOUR_OF_DAY) + c.get(java.util.Calendar.MINUTE) / 60.0
}

/** Midnight at the top, clocking round to the right. Compose's zero angle is at 3 o'clock, hence the −90. */
internal fun hourAngleDegrees(hour: Double): Double = hour / 24.0 * 360.0 - 90.0

/** Sweep from [from] to [to] clockwise, always positive so an arc crossing midnight still draws. */
internal fun sweepHours(from: Double, to: Double): Double =
    ((to - from) % 24.0).let { if (it <= 0.0) it + 24.0 else it }

/**
 * Rounded to five minutes: the underlying phase is an activity fit, so a to-the-minute caption would imply
 * a precision the estimate does not carry.
 */
@Composable
private fun alignmentText(offsetHours: Double): String {
    val minutes = (offsetHours * 60 / 5).roundToInt() * 5
    if (kotlin.math.abs(minutes) < 30) return uiString(R.string.l10n_body_clock_dial_card_in_sync_with_your_body_clock_4b5181c8)
    val absMinutes = kotlin.math.abs(minutes)
    val amount = if (absMinutes >= 60) String.format(java.util.Locale.getDefault(), "%.1f h", absMinutes / 60.0) else "$absMinutes min"
    return if (minutes > 0) uiString(R.string.l10n_body_clock_dial_card_1_s_later_than_your_body_50b7d966, amount) else uiString(R.string.l10n_body_clock_dial_card_1_s_earlier_than_your_body_68877f50, amount)
}

@Composable
private fun chronotypeText(c: CircadianEngine.Chronotype): String = when (c) {
    CircadianEngine.Chronotype.MORNING -> uiString(R.string.l10n_body_clock_dial_card_morning_type_289dd7bf)
    CircadianEngine.Chronotype.INTERMEDIATE -> uiString(R.string.l10n_body_clock_dial_card_intermediate_type_a7b6b74c)
    CircadianEngine.Chronotype.EVENING -> uiString(R.string.l10n_body_clock_dial_card_evening_type_3e1ce111)
}

File: Strand/Screens/BodyClockDialCard.swift
import SwiftUI
import StrandAnalytics
import StrandDesign

/// A 24 h dial comparing the night actually slept against the chronotype-ideal window (#1680).
///
/// Two concentric arcs on one ring: the outer is where the body clock wanted the night, the inner is
/// where it happened. Overlap is the whole message — the card exists so "was last night's timing right"
/// is answerable at a glance, which the existing text-only `BodyClockCard` on Health cannot do.
///
/// VOCABULARY. The caption measures `sleepWindowOffsetHours` — the distance between the two ARCS DRAWN —
/// and NOT `offsetVsScheduleMinutes`, which compares the clock to the wearer's habitual schedule and is
/// what the Health card already reports. The two disagree exactly when someone keeps a consistent
/// schedule that does not suit their clock, and that is the case this dial exists to show, so captioning
/// it with the other number would contradict the picture.
///
/// Nothing here computes a metric: the window, the offset and the chronotype all come from
/// `CircadianEngine`, byte-identical with the Kotlin twin. Only the drawing is per-platform.
struct BodyClockDialCard: View {
    let estimate: CircadianEngine.PhaseEstimate
    /// The night's own bed/wake clock hours (0..<24, fractional), from the scored session.
    let actualBedHour: Double
    let actualWakeHour: Double

    private var hue: Color { StrandPalette.restColor }

    /// The night's length, taken the long way round the clock when it crosses midnight.
    private var durationHours: Double {
        let d = (actualWakeHour - actualBedHour).truncatingRemainder(dividingBy: 24)
        return d <= 0 ? d + 24 : d
    }

    private var ideal: (bedHour: Double, wakeHour: Double)? {
        CircadianEngine.idealSleepWindow(tempMinHour: estimate.tempMinHour, durationHours: durationHours)
    }

    private var offsetHours: Double {
        CircadianEngine.sleepWindowOffsetHours(tempMinHour: estimate.tempMinHour,
                                               actualWakeHour: actualWakeHour)
    }

    var body: some View {
        // `ideal` is nil exactly when the night's length is non-positive or a full day — the same input
        // that makes `sweep` wrap to 24 h and draw the actual arc as a complete ring. Rendering a full
        // circle with no ideal arc beside it would state something false about the night, so the card
        // stands down instead. Twin of the Kotlin guard.
        if ideal != nil {
            NoopCard(tint: hue) {
                VStack(alignment: .leading, spacing: NoopMetrics.gap) {
                    header
                    dial
                        .frame(maxWidth: .infinity)
                        .accessibilityElement(children: .ignore)
                        // Label only. The verdict is the caption Text below, a separate element, so giving
                        // the dial the same string as its VALUE made VoiceOver announce it twice.
                        .accessibilityLabel(Text("Body clock dial"))
                    Text(alignmentText)
                        .font(StrandFont.title2)
                        .foregroundStyle(StrandPalette.textPrimary)
                        .fixedSize(horizontal: false, vertical: true)
                    if let chronotype = CircadianEngine.chronotype(estimate) {
                        Text(chronotypeText(chronotype))
                            .font(StrandFont.footnote)
                            .foregroundStyle(StrandPalette.textTertiary)
                    }
                }
            }
            .accessibilityElement(children: .contain)
        }
    }

    private var header: some View {
        HStack(alignment: .firstTextBaseline) {
            VStack(alignment: .leading, spacing: 2) {
                Text("Body clock").strandOverline()
                Text("Last night against your clock")
                    .font(StrandFont.footnote)
                    .foregroundStyle(StrandPalette.textTertiary)
            }
            Spacer()
        }
    }

    /// Midnight at the top, clocking round to the right — the orientation every 24 h dial uses, so the
    /// ring reads without a legend. SwiftUI's zero angle is at 3 o'clock, hence the −90.
    private func angle(_ hour: Double) -> Angle { .degrees(hour / 24 * 360 - 90) }

    /// Sweep from `from` to `to` going clockwise, always positive so an arc crossing midnight still draws.
    private func sweep(_ from: Double, _ to: Double) -> Double {
        let d = (to - from).truncatingRemainder(dividingBy: 24)
        return d <= 0 ? d + 24 : d
    }

    private var dial: some View {
        Canvas { ctx, size in
            let side = min(size.width, size.height)
            let centre = CGPoint(x: size.width / 2, y: size.height / 2)
            let outer = side / 2 - 10
            let inner = outer - 16

            // The bare ring, so an empty dial still reads as a clock rather than a broken chart.
            ctx.stroke(Path(ellipseIn: CGRect(x: centre.x - outer, y: centre.y - outer,
                                              width: outer * 2, height: outer * 2)),
                       with: .color(StrandPalette.hairline), lineWidth: 1)

            // Six-hourly ticks: enough to orient the eye, few enough not to compete with the arcs.
            for tick in stride(from: 0.0, to: 24.0, by: 6.0) {
                let a = angle(tick).radians
                var p = Path()
                p.move(to: CGPoint(x: centre.x + cos(a) * (outer - 4), y: centre.y + sin(a) * (outer - 4)))
                p.addLine(to: CGPoint(x: centre.x + cos(a) * outer, y: centre.y + sin(a) * outer))
                ctx.stroke(p, with: .color(StrandPalette.textTertiary.opacity(0.5)), lineWidth: 1)
            }

            func arc(radius: CGFloat, from: Double, to: Double, colour: Color, width: CGFloat) {
                var p = Path()
                p.addArc(center: centre, radius: radius,
                         startAngle: angle(from),
                         endAngle: .degrees(angle(from).degrees + sweep(from, to) / 24 * 360),
                         clockwise: false)
                ctx.stroke(p, with: .color(colour), style: StrokeStyle(lineWidth: width, lineCap: .round))
            }

            if let ideal {
                arc(radius: outer, from: ideal.bedHour, to: ideal.wakeHour,
                    colour: hue.opacity(0.35), width: 8)
            }
            arc(radius: inner, from: actualBedHour, to: actualWakeHour, colour: hue, width: 8)
        }
        .frame(height: 170)
    }

    /// Rounded to five minutes: the underlying phase is an activity fit, so a to-the-minute caption would
    /// imply a precision the estimate does not carry.
    private var alignmentText: String {
        let minutes = Int((offsetHours * 60 / 5).rounded()) * 5
        if abs(minutes) < 30 { return String(localized: "In sync with your body clock") }
        let hours = abs(Double(minutes)) / 60
        // Locale-formatted, NOT String(format:). That is C-locale, so it prints "1.5 h" for a German
        // reader while the Android twin's String.format prints "1,5 h" from the default locale — the two
        // platforms disagreeing on a decimal separator in the same sentence.
        let amount = hours >= 1
            ? "\(hours.formatted(.number.precision(.fractionLength(1)))) h"
            : "\(abs(minutes)) min"
        return minutes > 0
            ? String(localized: "\(amount) later than your body clock")
            : String(localized: "\(amount) earlier than your body clock")
    }

    private func chronotypeText(_ c: CircadianEngine.Chronotype) -> String {
        switch c {
        case .morning:      return String(localized: "Morning type")
        case .intermediate: return String(localized: "Intermediate type")
        case .evening:      return String(localized: "Evening type")
        }
    }
}
